File: src/main/java/com/example/usercrud/persistance/repository/UserRepositoryJDBCImpl.java
package com.example.usercrud.persistance.repository;

import com.example.usercrud.business.entity.User;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@RequiredArgsConstructor
public class UserRepositoryJDBCImpl implements UserRepository {
    private final Connection connection;

    @Override
    public Optional<User> findByEmail(String email) {
        Optional<User> user = Optional.empty();
        try {
            PreparedStatement statement = connection.prepareStatement("SELECT * FROM users WHERE email=?");
            statement.setString(1, email);
            ResultSet resultSet = statement.executeQuery();
            if(resultSet.next()) {
                user = Optional.of(new User(resultSet.getLong(1), resultSet.getString(2), resultSet.getString(3), resultSet.getString(4), resultSet.getString(5), resultSet.getInt(6), resultSet.getString(7), resultSet.getString(8)));
            }
            statement.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return user;
    }

    @Override
    public User getByEmail(String email) {
        User user = null;
        try {
            PreparedStatement statement = connection.prepareStatement("SELECT * FROM users WHERE email=?");
            statement.setString(1, email);
            ResultSet resultSet = statement.executeQuery();
            if(resultSet.next()) {
                user = new User(resultSet.getLong(1), resultSet.getString(2), resultSet.getString(3), resultSet.getString(4), resultSet.getString(5), resultSet.getInt(6), resultSet.getString(7), resultSet.getString(8));
            }
            statement.close();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return user;
    }

    @Override
    public boolean existsByEmail(String email) {
        boolean exists = false;
        try {
            PreparedStatement statement = connection.prepareStatement("SELECT * FROM users WHERE email=?");
            statement.setString(1, email);
            ResultSet resultSet = statement.executeQuery();
            exists = resultSet.next();
            statement.close();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return exists;
    }

    @Override
    public void deleteByEmail(String email) {
        try {
            PreparedStatement statement = connection.prepareStatement("DELETE * FROM users WHERE email=?");
            statement.setString(1, email);
            statement.executeUpdate();
            statement.close();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public List<User> findAllByCountry(String country) {
        ArrayList<User> users = new ArrayList<>();
        try {
            PreparedStatement statement = connection.prepareStatement("SELECT * FROM users WHERE country=?");
            statement.setString(1, country);
            ResultSet resultSet = statement.executeQuery();
            while(resultSet.next()) {
                users.add(new User(resultSet.getLong(1), resultSet.getString(2), resultSet.getString(3), resultSet.getString(4), resultSet.getString(5), resultSet.getInt(6), resultSet.getString(7), resultSet.getString(8)));
            }
            statement.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return users;
    }

    @Override
    public List<User> findAll() {
        ArrayList<User> users = new ArrayList<>();
        try (Statement statement = connection.createStatement()) {
            ResultSet resultSet = statement.executeQuery("SELECT * FROM users");
            while(resultSet.next()) {
                users.add(new User(resultSet.getLong(1), resultSet.getString(2), resultSet.getString(3), resultSet.getString(4), resultSet.getString(5), resultSet.getInt(6), resultSet.getString(7), resultSet.getString(8)));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return users;
    }

    @Override
    public User save(User user) {
        User savedUser = null;
        try {
            PreparedStatement statement = connection.prepareStatement("INSERT INTO users (first_name, last_name, middle_name, country, gender, email, avatar) VALUES(?, ?, ?, ?, ?, ?, ?)");
            statement.setString(1, user.getFirstName());
            statement.setString(2, user.getLastName());
            statement.setString(3, user.getMiddleName());
            statement.setString(4, user.getCountry());
            statement.setInt(5, user.getGender());
            statement.setString(6, user.getEmail());
            statement.setString(7, user.getAvatar());
            statement.executeUpdate();
            savedUser = getByEmail(user.getEmail());
            statement.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return savedUser;
    }

    @Override
    public Optional<User> findById(Long id) {
        Optional<User> user = Optional.empty();
        try {
            PreparedStatement statement = connection.prepareStatement("SELECT * FROM users WHERE id=?");
            statement.setLong(1, id);
            ResultSet resultSet = statement.executeQuery();
            if(resultSet.next()) {
                user = Optional.of(new User(resultSet.getLong(1), resultSet.getString(2), resultSet.getString(3), resultSet.getString(4), resultSet.getString(5), resultSet.getInt(6), resultSet.getString(7), resultSet.getString(8)));
            }
            statement.close();
        } catch(Exception e) {
            e.printStackTrace();
        }
        return user;
    }

    @Override
    public void deleteById(Long id) {
        try {
            PreparedStatement statement = connection.prepareStatement("DELETE * FROM users WHERE id=?");
            statement.setLong(1, id);
            statement.executeUpdate();
            statement.close();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean existsById(Long id) {
        boolean exists = false;
        try {
            PreparedStatement statement = connection.prepareStatement("SELECT * FROM users WHERE id=?");
            statement.setLong(1, id);
            ResultSet resultSet = statement.executeQuery();
            exists = resultSet.next();
            statement.close();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return exists;
    }
}

File: src/main/kotlin/cloud/folium/usercrud/persistance/repository/UserRepositoryJDBCImpl.kt
package cloud.folium.usercrud.persistance.repository

import cloud.folium.usercrud.business.entity.User
import kotlin.Throws
import org.springframework.stereotype.Repository
import java.lang.Exception
import java.sql.Connection
import java.sql.SQLException
import java.util.*

class UserRepositoryJDBCImpl : UserRepository {
    private val connection: Connection? = null
    @Throws(SQLException::class)
    override fun findByEmail(email: String?): User? {
        var user: User? = null
        try {
            connection!!.prepareStatement("SELECT * FROM users WHERE email=?").use { statement ->
                statement.setString(1, email)
                val resultSet = statement.executeQuery()
                if (resultSet.next()) {
                    user =
                        User(
                            resultSet.getLong(1),
                            resultSet.getString(2),
                            resultSet.getString(3),
                            resultSet.getString(4),
                            resultSet.getString(5),
                            resultSet.getInt(6),
                            resultSet.getString(7),
                            resultSet.getString(8)
                        )
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            throw e
        }
        return user
    }

    @Throws(SQLException::class)
    override fun getByEmail(email: String?): User? {
        var user: User? = null
        try {
            connection!!.prepareStatement("SELECT * FROM users WHERE email=?").use { statement ->
                statement.setString(1, email)
                val resultSet = statement.executeQuery()
                if (resultSet.next()) {
                    user = User(
                        resultSet.getLong(1),
                        resultSet.getString(2),
                        resultSet.getString(3),
                        resultSet.getString(4),
                        resultSet.getString(5),
                        resultSet.getInt(6),
                        resultSet.getString(7),
                        resultSet.getString(8)
                    )
                }
            }
        } catch (e: SQLException) {
            e.printStackTrace()
            throw e
        }
        return user
    }

    @Throws(SQLException::class)
    override fun existsByEmail(email: String?): Boolean {
        var exists = false
        try {
            connection!!.prepareStatement("SELECT * FROM users WHERE email=?").use { statement ->
                statement.setString(1, email)
                exists = statement.execute()
            }
        } catch (e: SQLException) {
            e.printStackTrace()
            throw e
        }
        return exists
    }

    @Throws(SQLException::class)
    override fun deleteByEmail(email: String?) {
        try {
            connection!!.prepareStatement("DELETE * FROM users WHERE email=?").use { statement ->
                statement.setString(1, email)
                statement.executeUpdate()
            }
        } catch (e: SQLException) {
            e.printStackTrace()
            throw e
        }
    }

    @Throws(SQLException::class)
    override fun findAllByCountry(country: String?): List<User> {
        val users = ArrayList<User>()
        try {
            connection!!.prepareStatement("SELECT * FROM users WHERE country=?").use { statement ->
                statement.setString(1, country)
                val resultSet = statement.executeQuery()
                while (resultSet.next()) {
                    users.add(
                        User(
                            resultSet.getLong(1),
                            resultSet.getString(2),
                            resultSet.getString(3),
                            resultSet.getString(4),
                            resultSet.getString(5),
                            resultSet.getInt(6),
                            resultSet.getString(7),
                            resultSet.getString(8)
                        )
                    )
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            throw e
        }
        return users
    }

    @Throws(SQLException::class)
    override fun findAll(): List<User> {
        val users = ArrayList<User>()
        try {
            connection!!.createStatement().use { statement ->
                val resultSet = statement.executeQuery("SELECT * FROM users")
                while (resultSet.next()) {
                    users.add(
                        User(
                            resultSet.getLong(1),
                            resultSet.getString(2),
                            resultSet.getString(3),
                            resultSet.getString(4),
                            resultSet.getString(5),
                            resultSet.getInt(6),
                            resultSet.getString(7),
                            resultSet.getString(8)
                        )
                    )
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            throw e
        }
        return users
    }

    override fun save(user: User?): User? {
        var savedUser: User? = null
        try {
            connection!!.prepareStatement("INSERT INTO users (first_name, last_name, middle_name, country, gender, email, avatar) VALUES(?, ?, ?, ?, ?, ?, ?)")
                .use { statement ->
                    statement.setString(1, user?.firstName)
                    statement.setString(2, user?.lastName)
                    statement.setString(3, user?.middleName)
                    statement.setString(4, user?.country)
                    // todo
                    user?.gender?.let { statement.setInt(5, it) }
                    statement.setString(6, user?.email)
                    statement.setString(7, user?.avatar)
                    statement.executeUpdate()
                    savedUser = getByEmail(user?.email)
                }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return savedUser
    }

    @Throws(SQLException::class)
    override fun findById(id: Long): User? {
        var user: User? = null
        try {
            connection!!.prepareStatement("SELECT * FROM users WHERE id=?").use { statement ->
                statement.setLong(1, id)
                val resultSet = statement.executeQuery()
                if (resultSet.next()) {
                    user =
                        User(
                            resultSet.getLong(1),
                            resultSet.getString(2),
                            resultSet.getString(3),
                            resultSet.getString(4),
                            resultSet.getString(5),
                            resultSet.getInt(6),
                            resultSet.getString(7),
                            resultSet.getString(8)
                        )
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            throw e
        }
        return user
    }

    @Throws(SQLException::class)
    override fun deleteById(id: Long) {
        try {
            connection!!.prepareStatement("DELETE * FROM users WHERE id=?").use { statement ->
                statement.setLong(1, id)
                statement.executeUpdate()
            }
        } catch (e: SQLException) {
            e.printStackTrace()
            throw e
        }
    }

    @Throws(SQLException::class)
    override fun existsById(id: Long): Boolean {
        var exists = try {
            connection!!.prepareStatement("SELECT * FROM users WHERE id=?").use { statement ->
                statement.setLong(1, id)
                statement.execute()
            }
        } catch (e: SQLException) {
            e.printStackTrace()
            throw e
        }
        return exists
    }
}